cmd/query/app/server_test.go

```diff
@@ -100,17 +100,61 @@ func TestServer(t *testing.T) {
 	assert.Equal(t, healthcheck.Unavailable, flagsSvc.HC().Get())
 }
 
+func TestServer2(t *testing.T) {
+	flagsSvc := flags.NewService(ports.QueryAdminHTTP)
+	flagsSvc.Logger = zap.NewNop()
+	// hostPort := ports.GetAddressFromCLIOptions(ports.QueryHTTP, "")
+
+	spanReader := &spanstoremocks.Reader{}
+	dependencyReader := &depsmocks.Reader{}
+	expectedServices := []string{"test"}
+	spanReader.On("GetServices", mock.AnythingOfType("*context.valueCtx")).Return(expectedServices, nil)
+
+	querySvc := querysvc.NewQueryService(spanReader, dependencyReader, querysvc.QueryServiceOptions{})
+
+	server, err := NewServer(flagsSvc.Logger, querySvc,
+		&QueryOptions{HTTPHostPort: "127.0.0.1:8080", GRPCHostPort: "127.0.0.1:8081", BearerTokenPropagation: true},
+		opentracing.NoopTracer{})
+	assert.Nil(t, err)
+	assert.NoError(t, server.Start())
+	go func() {
+		for s := range server.HealthCheckStatus() {
+			flagsSvc.SetHealthCheckStatus(s)
+		}
+	}()
+
+	client := newGRPCClient(t, "127.0.0.1:8081")
+	defer client.conn.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+
+	res, err := client.GetServices(ctx, &api_v2.GetServicesRequest{})
+	assert.NoError(t, err)
+	assert.Equal(t, expectedServices, res.Services)
+
+	server.Close()
+	for i := 0; i < 10; i++ {
+		if flagsSvc.HC().Get() == healthcheck.Unavailable {
```

Comment:
I wonder if we have a new flaky test here (I don't remember if the `TestServer` was flaky), but ideally, you'd create a `wg := sync.WaitGroup{}` at the beginning, with `wg.Add(1)`. In the line where you set the health check status, you'd check if it contains the state you are expecting and then call `wg.Done()`. At the end of your test, you'd call `wg.Wait()`. If the test times out, your test failed.